consensus/hotstuff/forks/forks2.go

```diff
@@ -0,0 +1,557 @@
+// block is the head of a 2-chain satisfying the finalization rule, we update `Forks.finalityProof` to
+// the new latest finalized block. Calling this method with previously-processed blocks leaves the
+// consensus state invariant.
+// UNVALIDATED: assumes that relevant block properties are consistent with previous blocks
+// Error returns:
+//   - model.MissingBlockError if the parent does not exist in the forest (but is above
+//     the pruned view). From the perspective of Forks, this error is benign (no-op).
+//   - model.ByzantineThresholdExceededError in case observing a finalization fork (violating
+//     a foundational consensus guarantee). This indicates that there are 1/3+ Byzantine nodes
+//     (weighted by stake) in the network, breaking the safety guarantees of HotStuff (or there
+//     is a critical bug / data corruption). Forks cannot recover from this exception.
+//   - generic error in case of unexpected bug or internal state corruption
+func (f *Forks2) checkForAdvancingFinalization(certifiedBlock *model.CertifiedBlock) error {
+	// We prune all blocks in forest which are below the most recently finalized block.
+	// Hence, we have a pruned ancestry if and only if either of the following conditions applies:
+	//    (a) If a block's parent view (i.e. block.QC.View) is below the most recently finalized block.
+	//    (b) If a block's view is equal to the most recently finalized block.
+	// Caution:
+	// * Under normal operation, case (b) is covered by the logic for case (a)
+	// * However, the existence of a genesis block requires handling case (b) explicitly:
+	//   The root block is specified and trusted by the node operator. If the root block is the
+	//   genesis block, it might not contain a QC pointing to a parent (as there is no parent).
+	//   In this case, condition (a) cannot be evaluated.
+	lastFinalizedView := f.FinalizedView()
+	if (certifiedBlock.View() <= lastFinalizedView) || (certifiedBlock.Block.QC.View < lastFinalizedView) {
+		// Repeated blocks are expected during normal operations. We enter this code block if and only
+		// if the parent's view is _below_ the last finalized block. It is straight forward to show:
+		// Lemma: Let B be a block whose 2-chain reaches beyond the last finalized block
+		//        => B will not update the locked or finalized block
+		return nil
+	}
+
+	// retrieve parent; always expected to succeed, because we passed the checks above
+	qcForParent := certifiedBlock.Block.QC
+	parentVertex, parentBlockKnown := f.forest.GetVertex(qcForParent.BlockID)
+	if !parentBlockKnown {
+		return model.MissingBlockError{View: qcForParent.View, BlockID: qcForParent.BlockID}
+	}
+	parentBlock := parentVertex.(*BlockContainer2).Block()
+
+	// Note: we assume that all stored blocks pass Forks.EnsureBlockIsValidExtension(block);
+	//       specifically, that Proposal's ViewNumber is strictly monotonously
```

Comment:
```suggestion
	//       specifically, that Proposal's ViewNumber is strictly monotonically
```

Reply:
Fixed the incorrect usages within Forks 👉  https://github.com/onflow/flow-go/pull/4190/commits/0b531f40e749c12d206236dc196b2e6183f482b2